core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java

```diff
@@ -1381,6 +1424,58 @@ private static void adjustTypeForMultisetConstructor(
     }
   }
 
+  /**
+   * Adjusts the types for operands in a SqlCallBinding during the construction of a sql collection
+   * type such as array_append and array_prepend, but cannot change the data type in the array.
+   *
+   * @param evenType the {@link RelDataType} to which the operands at even positions should be cast
+   * @param oddType the {@link RelDataType} to which the operands at odd positions should be cast
+   * @param sqlCallBinding the {@link SqlCallBinding} containing the operands to be adjusted
+   */
+  private static void adjustTypeForMultisetConstructor2(
```

Comment:
I'm thinking about whether adjustTypeForMultisetConstructor2 and adjustTypeForMultisetConstructor3 can be combined into one function. At present, I don't have a good solution, so I separate the two functions.